### ISSUE TITLE
THP can cause a severe performance issue on Linux (x86) Cloud Foundry environment

### ISSUE THREAD

**Issue by klangman** [eclipse/openj9#8546]
A OpenJ9 consumer has encountered a severe performance issue after Transparent Huge Pages was enabled by default on Linux X86. #6713

A profile from the effected system showed that the kernel routine "list_lru_count_one" was taking 95% of the CPU time. With a bit of google'ing I found the following kernel fix that I suspect addresses the performance issue.

https://lore.kernel.org/lkml/153063052519.1818.9393587113056959488.stgit@localhost.localdomain/T/#m41c9eab50fc96f80572c6e2a605532eb6154d431

Testing showed that THP was triggering the issue to occur and they are now happily using: 
"-XX:-TransparentHugePage"

It might be worth investigating if we should only be enabling THP by default when executing on a kernel which is not effected by this perf issue.

**Comment by DanHeidinga** [comment]
> It might be worth investigating if we should only be enabling THP by default when executing on a kernel which is not effected by this perf issue.

@klangman Do you have the range of kernels that are affected / not affected by this issue?

**Comment by klangman** [comment]
@DanHeidinga According to the documentation that I found, the issues was fixed in 4.19. When it was introduced, I have no idea other to say that the issue is attached to cgroups in some say.  Maybe we should just disable THP on small footprint builds when using a kernel older then 4.19 or maybe there is a way to know when cgroups are being used?

**Comment by dmitripivkine** [comment]
>... maybe there is a way to know when cgroups are being used?

GC code has this check to make special settings for run in container: https://github.com/eclipse/openj9/blob/041daeec2e693cc6ac6107f2b3c9f71db319b0eb/runtime/gc_base/GCExtensions.cpp#L244

**Comment by DanHeidinga** [comment]
Reading Kevin's link, the issue manifests on systems with many cgroups which would imply that it can happen whether or no the VM is running in a cgroup.

We'd need to better understand the conditions that cause the issue before we enable / disable this.  One data point is hard to generalize from =)

**Comment by dmitripivkine** [comment]
> Reading Kevin's link, the issue manifests on systems with many cgroups which would imply that it can happen whether or no the VM is running in a cgroup.
> 
> We'd need to better understand the conditions that cause the issue before we enable / disable this. One data point is hard to generalize from =)

Ok, thank you. I misunderstood this. This means there is not so much we can do about this runtime - well may be an option to disable usage of THP in Java as workaround (but would it more correct to disable THP on OS level by changing `transparent_hugepage=never`):
```
# cat /sys/kernel/mm/transparent_hugepage/enabled
[always] madvise never
```
Most likely slow handling of THP relates with general memory fragmentation when OS needs to reallocate a lot of data to get contiguous huge page. This memory fragmentation might be caused by apps running (containers for example) but it is out of JVM functionality scope obviously

**Comment by pshipton** [comment]
It doesn't seem this will be resolved for the 0.20 release, moving it out.

**Comment by DanHeidinga** [comment]
I don't think this is something the JVM can address.  THP enabled + lots of cgroups on the same machine with particular kernel versions is an environment issue outside the vm's purview. 

It's also a dynamic condition - we'd have to monitor the system to see if the number of cgroups grew past some threshold.

I'm going to close this based on the above
